sklego/model_selection.py

```diff
@@ -6,76 +6,151 @@
         self.valid_duration = valid_duration
         self.gap_duration = gap_duration
 
+    def join_date_and_x(self, X):
+        """
+        Make a DataFrame indexed by the pandas index (the same as date_series) with date column joined with that index
+        and with the 'numpy index' column (i.e. just a range) that is required for the output and the rest of sklearn
+        :param pandas.DataFrame X:
+        """
+        X_index_df = pd.DataFrame(range(len(X)), columns=['np_index'], index=X.index)
+        X_index_df = X_index_df.join(self.date_serie)
+
+        return X_index_df
+
     def split(self, X, y=None, groups=None):
         """
         Generate indices to split data into training and test set.
         :param pandas.DataFrame X:
         :param y: Always ignored, exists for compatibility
         :param groups: Always ignored, exists for compatibility
         """
-        date_series = self.df.loc[X.index][self.date_col]
-        date_series = date_series.sort_values(ascending=True)
-        date_min = date_series.min()
-        date_max = date_series.max()
+
+        X_index_df = self.join_date_and_x(X)
+        X_index_df = X_index_df.sort_values('__date__', ascending=True)
+
+        if len(X) != len(X_index_df):
+            raise AssertionError("X and X_index_df are not the same lenght, "
+                                 "there must be some index missing in 'self.date_serie'")
+
+        date_min = X_index_df['__date__'].min()
+        date_max = X_index_df['__date__'].max()
 
         current_date = date_min
         while True:
             if current_date + self.train_duration + self.valid_duration > date_max:
                 break
 
-            train_i = date_series[
-                (date_series >= current_date) &
-                (date_series < current_date + self.train_duration - self.gap_duration)].index.values
-            valid_i = date_series[
-                (date_series >= current_date + self.train_duration) &
-                (date_series < current_date + self.train_duration + self.valid_duration)].index.values
+            X_train_df = X_index_df[
+                (X_index_df['__date__'] >= current_date) &
+                (X_index_df['__date__'] < current_date + self.train_duration - self.gap_duration)]
+            X_valid_df = X_index_df[
+                (X_index_df['__date__'] >= current_date + self.train_duration) &
+                (X_index_df['__date__'] < current_date + self.train_duration + self.valid_duration)]
 
             current_date = current_date + self.valid_duration
 
-            yield (np.array([X.index.get_loc(i) for i in train_i]),
-                   np.array([X.index.get_loc(i) for i in valid_i]))
+            yield (X_train_df['np_index'].values,
+                   X_valid_df['np_index'].values)
 
     def get_n_splits(self, X=None, y=None, groups=None):
 
         return sum(1 for x in self.split(X, y, groups))
 
+    def plot(self, X):
```

Comment:
I'm not sure whether this should be on the class. I think it'd be nicer to just have it in the documentation

Reply:
I think it is really useful for the DS to be able how was the split done.